crypto/bls12/bls12.go

```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+func (bls *bls12Base) aggregateVerify(publicKeys []*PublicKey, messages [][]byte, signature *bls12.PointG2) bool {
+	set := make(map[string]struct{})
+	for _, m := range messages {
+		set[string(m)] = struct{}{}
+	}
+	return len(messages) == len(set) && bls.coreAggregateVerify(publicKeys, messages, signature)
+}
+
+func (bls *bls12Base) fastAggregateVerify(publicKeys []*PublicKey, message []byte, signature *bls12.PointG2) bool {
+	engine := bls12.NewEngine()
+	var aggregate bls12.PointG1
+	for _, pk := range publicKeys {
+		engine.G1.Add(&aggregate, &aggregate, pk.p)
+	}
+	return bls.coreVerify(&PublicKey{p: &aggregate}, message, signature, domain)
+}
+
+// Sign creates a cryptographic signature of the given messsage.
+func (bls *bls12Base) Sign(message []byte) (signature consensus.QuorumSignature, err error) {
+	// TODO: consider passing message to this function instead of hash
+	p, err := bls.coreSign(message, domain)
+	if err != nil {
+		return nil, fmt.Errorf("bls12: coreSign failed: %w", err)
+	}
+	bf := crypto.Bitfield{}
+	bf.Add(bls.mods.ID())
+	return &AggregateSignature{sig: *p, participants: bf}, nil
+}
+
+// Verify verifies the given cryptographic signature according to the specified options.
+// NOTE: One of either VerifySingle or VerifyMulti options MUST be specified,
+// otherwise this function will have nothing to verify the signature against.
+func (bls *bls12Base) Verify(signature consensus.QuorumSignature, options ...consensus.VerifyOption) bool {
 	sig, ok := signature.(*AggregateSignature)
 	if !ok {
+		panic(fmt.Sprintf("cannot verify signature of incompatible type %T (expected %T)", signature, sig))
+	}
+
+	var opts consensus.VerifyOptions
+	for _, opt := range options {
+		opt(&opts)
+	}
+
+	if !opts.MultipleMessages && opts.Message == nil {
+		panic("no hash(es) to verify the signature against: you must specify one of the VerifyHash or VerifyHashes options")
+	}
+
+	l := sig.Participants().Len()
+
+	if l < opts.Threshold {
 		return false
 	}
-	hashSet := make(map[consensus.Hash]struct{})
-	engine := bls12.NewEngine()
-	engine.AddPairInv(&bls12.G1One, &sig.sig)
-	for id, hash := range hashes {
-		if _, ok := hashSet[hash]; ok {
-			continue
-		}
-		hashSet[hash] = struct{}{}
-		replica, ok := bc.mods.Configuration().Replica(id)
+
+	var (
+		publicKeys = make([]*PublicKey, 0, l)
+		messages   = make([][]byte, 0, l)
+		valid      = true
+	)
+
+	// get all public keys and messages to verify
+	sig.participants.RangeWhile(func(i hotstuff.ID) bool {
+		replica, ok := bls.mods.Configuration().Replica(i)
 		if !ok {
+			valid = false
 			return false
 		}
-		pk, ok := replica.PublicKey().(*PublicKey)
-		if !ok {
+
+		if replica.ID() != bls.mods.ID() && !bls.checkPop(replica) {
+			valid = false
 			return false
 		}
-		p2, err := bls12.NewG2().HashToCurve(hash[:], domain)
-		if err != nil {
-			return false
+		publicKeys = append(publicKeys, replica.PublicKey().(*PublicKey))
+
+		if opts.MultipleMessages {
```

Comment:
If `MultipleMessages` is false, then `messages` above is not used.

Overall, it seems like the code could be simplified by avoiding the separation between single message verify functions and multi-message verify functions.

Reply:
The problem with removing this separation, i.e. using the map for both `VerifySingle` and `VerifyMulti` is that it becomes difficult to check which option was used. For example, if the length is 1 and there is only one participant in the signature, then you still don't know if `VerifySingle` or `VerifyMulti` was used.

Reply:
Okay.